// @art-design-digital/payload-plugin-teammembers/dist/collections/Departments.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Departments = void 0;
var departments_1 = require("../translations/departments");
var types_1 = require("../types");
var deepmerge_1 = __importDefault(require("../utils/deepmerge"));
var Departments = function (userPluginOptions) {
    var pluginOptions = (0, deepmerge_1.default)(types_1.defaultPluginOptions, userPluginOptions);
    return {
        slug: 'departments',
        admin: {
            group: (pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.adminGroup) || undefined,
            useAsTitle: 'name',
            defaultColumns: ['name'],
            description: departments_1.translations.admin.description,
        },
        labels: {
            singular: departments_1.translations.labels.singular,
            plural: departments_1.translations.labels.plural,
        },
        access: {
            create: function () { return true; },
            read: function () { return true; },
            update: function () { return true; },
            delete: function () { return true; },
        },
        fields: [
            {
                type: 'text',
                name: 'name',
                label: departments_1.translations.fields.name.label,
                localized: pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.localized,
                required: true,
                admin: {
                    description: departments_1.translations.fields.name.admin.description,
                },
            },
        ],
    };
};
exports.Departments = Departments;
