// payload/dist/admin/components/elements/Collapsible/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Collapsible", {
    enumerable: true,
    get: function() {
        return Collapsible;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reactanimateheight = /*#__PURE__*/ _interop_require_default(require("react-animate-height"));
const _reacti18next = require("react-i18next");
const _Chevron = /*#__PURE__*/ _interop_require_default(require("../../icons/Chevron"));
const _Drag = /*#__PURE__*/ _interop_require_default(require("../../icons/Drag"));
require("./index.scss");
const _provider = require("./provider");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'collapsible';
const Collapsible = ({ actions, children, className, collapsed: collapsedFromProps, collapsibleStyle = 'default', dragHandleProps, header, initCollapsed, onToggle })=>{
    const [collapsedLocal, setCollapsedLocal] = (0, _react.useState)(Boolean(initCollapsed));
    const [hoveringToggle, setHoveringToggle] = (0, _react.useState)(false);
    const { withinCollapsible } = (0, _provider.useCollapsible)();
    const { t } = (0, _reacti18next.useTranslation)('fields');
    const collapsed = typeof collapsedFromProps === 'boolean' ? collapsedFromProps : collapsedLocal;
    const toggleCollapsible = _react.default.useCallback(()=>{
        if (typeof onToggle === 'function') onToggle(!collapsed);
        setCollapsedLocal(!collapsed);
    }, [
        onToggle,
        collapsed
    ]);
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            baseClass,
            className,
            dragHandleProps && `${baseClass}--has-drag-handle`,
            collapsed && `${baseClass}--collapsed`,
            withinCollapsible && `${baseClass}--nested`,
            hoveringToggle && `${baseClass}--hovered`,
            `${baseClass}--style-${collapsibleStyle}`
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_provider.CollapsibleProvider, {
        collapsed: collapsed,
        toggle: toggleCollapsible
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__toggle-wrap`,
        onMouseEnter: ()=>setHoveringToggle(true),
        onMouseLeave: ()=>setHoveringToggle(false)
    }, dragHandleProps && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__drag`,
        ...dragHandleProps.attributes,
        ...dragHandleProps.listeners
    }, /*#__PURE__*/ _react.default.createElement(_Drag.default, null)), /*#__PURE__*/ _react.default.createElement("button", {
        className: [
            `${baseClass}__toggle`,
            `${baseClass}__toggle--${collapsed ? 'collapsed' : 'open'}`
        ].filter(Boolean).join(' '),
        onClick: toggleCollapsible,
        type: "button"
    }, /*#__PURE__*/ _react.default.createElement("span", null, t('toggleBlock'))), header && /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            `${baseClass}__header-wrap`,
            dragHandleProps && `${baseClass}__header-wrap--has-drag-handle`
        ].filter(Boolean).join(' ')
    }, header && header), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__actions-wrap`
    }, actions && /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__actions`
    }, actions), /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__indicator`
    }, /*#__PURE__*/ _react.default.createElement(_Chevron.default, {
        direction: !collapsed ? 'up' : undefined
    })))), /*#__PURE__*/ _react.default.createElement(_reactanimateheight.default, {
        duration: 200,
        height: collapsed ? 0 : 'auto'
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__content`
    }, children))));
};
