// payload/dist/exports/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    afterReadPromise: function() {
        return _promise.promise;
    },
    afterReadTraverseFields: function() {
        return _traverseFields.traverseFields;
    },
    combineMerge: function() {
        return _combineMerge.combineMerge;
    },
    configToJSONSchema: function() {
        return _configToJSONSchema.configToJSONSchema;
    },
    createArrayFromCommaDelineated: function() {
        return _createArrayFromCommaDelineated.createArrayFromCommaDelineated;
    },
    deepCopyObject: function() {
        return _deepCopyObject.deepCopyObject;
    },
    deepMerge: function() {
        return _deepMerge.deepMerge;
    },
    entityToJSONSchema: function() {
        return _configToJSONSchema.entityToJSONSchema;
    },
    extractTranslations: function() {
        return _extractTranslations.extractTranslations;
    },
    fieldSchemaToJSON: function() {
        return _fieldSchemaToJSON.fieldSchemaToJSON;
    },
    fieldsToJSONSchema: function() {
        return _configToJSONSchema.fieldsToJSONSchema;
    },
    flattenTopLevelFields: function() {
        return _flattenTopLevelFields.default;
    },
    formatLabels: function() {
        return _formatLabels.formatLabels;
    },
    formatNames: function() {
        return _formatLabels.formatNames;
    },
    getCollectionIDFieldTypes: function() {
        return _getCollectionIDFieldTypes.getCollectionIDFieldTypes;
    },
    getIDType: function() {
        return _getIDType.getIDType;
    },
    getTranslation: function() {
        return _getTranslation.getTranslation;
    },
    i18nInit: function() {
        return _init.i18nInit;
    },
    isValidID: function() {
        return _isValidID.isValidID;
    },
    toWords: function() {
        return _formatLabels.toWords;
    },
    withMergedProps: function() {
        return _WithMergedProps.withMergedProps;
    },
    withNullableJSONSchemaType: function() {
        return _configToJSONSchema.withNullableJSONSchemaType;
    }
});
const _WithMergedProps = require("../admin/components/utilities/WithMergedProps");
const _promise = require("../fields/hooks/afterRead/promise");
const _traverseFields = require("../fields/hooks/afterRead/traverseFields");
const _extractTranslations = require("../translations/extractTranslations");
const _init = require("../translations/init");
const _combineMerge = require("../utilities/combineMerge");
const _configToJSONSchema = require("../utilities/configToJSONSchema");
const _createArrayFromCommaDelineated = require("../utilities/createArrayFromCommaDelineated");
const _deepCopyObject = require("../utilities/deepCopyObject");
const _deepMerge = require("../utilities/deepMerge");
const _fieldSchemaToJSON = require("../utilities/fieldSchemaToJSON");
const _flattenTopLevelFields = /*#__PURE__*/ _interop_require_default(require("../utilities/flattenTopLevelFields"));
const _formatLabels = require("../utilities/formatLabels");
const _getCollectionIDFieldTypes = require("../utilities/getCollectionIDFieldTypes");
const _getIDType = require("../utilities/getIDType");
const _getTranslation = require("../utilities/getTranslation");
const _isValidID = require("../utilities/isValidID");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
