// payload/dist/bin/generateTypes.js
/* eslint-disable @typescript-eslint/no-floating-promises */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "generateTypes", {
    enumerable: true,
    get: function() {
        return generateTypes;
    }
});
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _jsonschematotypescript = require("json-schema-to-typescript");
const _ = /*#__PURE__*/ _interop_require_default(require(".."));
const _load = /*#__PURE__*/ _interop_require_default(require("../config/load"));
const _configToJSONSchema = require("../utilities/configToJSONSchema");
const _logger = /*#__PURE__*/ _interop_require_default(require("../utilities/logger"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
async function generateTypes() {
    const logger = (0, _logger.default)();
    const config = await (0, _load.default)();
    const outputFile = process.env.PAYLOAD_TS_OUTPUT_PATH || config.typescript.outputFile;
    await _.default.init({
        disableDBConnect: true,
        disableOnInit: true,
        local: true,
        secret: '--unused--'
    });
    logger.info('Compiling TS types for Collections and Globals...');
    const jsonSchema = (0, _configToJSONSchema.configToJSONSchema)(_.default.config, _.default.db.defaultIDType, _.default);
    const declare = `declare module 'payload' {\n  export interface GeneratedTypes extends Config {}\n}`;
    (0, _jsonschematotypescript.compile)(jsonSchema, 'Config', {
        bannerComment: '/* tslint:disable */\n/* eslint-disable */\n/**\n* This file was automatically generated by Payload.\n* DO NOT MODIFY IT BY HAND. Instead, modify your source Payload config,\n* and re-run `payload generate:types` to regenerate this file.\n*/',
        style: {
            singleQuote: true
        },
        // Generates code for $defs that aren't referenced by the schema. Reason:
        // If a field defines an interfaceName, it should be included in the generated types
        // even if it's not used by another type. Reason: the user might want to use it in their own code.
        unreachableDefinitions: true
    }).then((compiled)=>{
        if (config.typescript.declare !== false) {
            compiled += `\n\n${declare}`;
        }
        _fs.default.writeFileSync(outputFile, compiled);
        logger.info(`Types written to ${outputFile}`);
    });
}
// when generateTypes.js is launched directly
if (module.id === require.main.id) {
    generateTypes();
}
