// payload/dist/admin/components/views/collections/Edit/Auth/APIKey.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacti18next = require("react-i18next");
const _uuid = require("uuid");
const _validations = require("../../../../../../fields/validations");
const _CopyToClipboard = /*#__PURE__*/ _interop_require_default(require("../../../../elements/CopyToClipboard"));
const _GenerateConfirmation = /*#__PURE__*/ _interop_require_default(require("../../../../elements/GenerateConfirmation"));
const _context = require("../../../../forms/Form/context");
const _Label = /*#__PURE__*/ _interop_require_default(require("../../../../forms/Label"));
const _shared = require("../../../../forms/field-types/shared");
const _useField = /*#__PURE__*/ _interop_require_default(require("../../../../forms/useField"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const path = 'apiKey';
const baseClass = 'api-key';
const APIKey = ({ enabled, readOnly })=>{
    const [initialAPIKey] = (0, _react.useState)((0, _uuid.v4)());
    const [highlightedField, setHighlightedField] = (0, _react.useState)(false);
    const { t } = (0, _reacti18next.useTranslation)();
    const apiKey = (0, _context.useFormFields)(([fields])=>fields[path]);
    const validate = (val)=>(0, _validations.text)(val, {
            data: {},
            maxLength: 48,
            minLength: 24,
            siblingData: {},
            t
        });
    const apiKeyValue = apiKey?.value;
    const APIKeyLabel = (0, _react.useMemo)(()=>/*#__PURE__*/ _react.default.createElement("div", {
            className: `${baseClass}__label`
        }, /*#__PURE__*/ _react.default.createElement("span", null, "API Key"), /*#__PURE__*/ _react.default.createElement(_CopyToClipboard.default, {
            value: apiKeyValue
        })), [
        apiKeyValue
    ]);
    const fieldType = (0, _useField.default)({
        path: 'apiKey',
        validate
    });
    const highlightField = ()=>{
        if (highlightedField) {
            setHighlightedField(false);
        }
        setTimeout(()=>{
            setHighlightedField(true);
        }, 1);
    };
    const { setValue, value } = fieldType;
    (0, _react.useEffect)(()=>{
        if (!apiKeyValue && enabled) {
            setValue(initialAPIKey);
        }
        if (!enabled) {
            setValue(null);
        }
    }, [
        apiKeyValue,
        enabled,
        setValue,
        initialAPIKey
    ]);
    (0, _react.useEffect)(()=>{
        if (highlightedField) {
            setTimeout(()=>{
                setHighlightedField(false);
            }, 10000);
        }
    }, [
        highlightedField
    ]);
    if (!enabled) {
        return null;
    }
    return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/ _react.default.createElement("div", {
        className: [
            _shared.fieldBaseClass,
            'api-key',
            'read-only'
        ].filter(Boolean).join(' ')
    }, /*#__PURE__*/ _react.default.createElement(_Label.default, {
        htmlFor: path,
        label: APIKeyLabel
    }), /*#__PURE__*/ _react.default.createElement("input", {
        className: highlightedField ? 'highlight' : undefined,
        disabled: true,
        id: "apiKey",
        name: "apiKey",
        type: "text",
        value: value || ''
    })), !readOnly && /*#__PURE__*/ _react.default.createElement(_GenerateConfirmation.default, {
        highlightField: highlightField,
        setKey: ()=>setValue((0, _uuid.v4)())
    }));
};
const _default = APIKey;
